### PR TITLE
Support non-UTC timezone for casting from date type to timestamp type [databricks]

### LINKED ISSUE #11461: [FEA] Support non-UTC timezone for casting from date to timestamp
**Is your feature request related to a problem? Please describe.**
From a customer, need support non-UTC timezone for casting from date to timestamp
Tests:
```
TZ=Asia/Shanghai ./integration_tests/run_pyspark_from_build.sh -s -k test_date_to_timestamp
```

case:
```python
@pytest.mark.parametrize('parser_policy', ["LEGACY", "CORRECTED", "EXCEPTION"], ids=idfn)
def test_date_to_timestamp(parser_policy):
    parser_policy_dic = {"spark.sql.legacy.timeParserPolicy": "{}".format(parser_policy)}
    incompatible_dic = {"spark.rapids.sql.incompatibleDateFormats.enabled": True}
    assert_gpu_and_cpu_are_equal_sql(
        lambda spark : unary_op_df(spark, date_gen),
        "tab",
        "SELECT cast(a as timestamp) from tab",
        conf=copy_and_update(parser_policy_dic, incompatible_dic))

```

### PR DESCRIPTION
**Pull request by res-life**
closes #11461

Only support non-DST(daylight saving time) time zones.

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
### non-DST timezone
passed:
```
TZ=Asia/Shanghai ./integration_tests/run_pyspark_from_build.sh -s -k test_date_to_timestamp
```

### For DST timezone:
Report not supporting prompt:

```
TZ=America/Los_Angeles ./integration_tests/run_pyspark_from_build.sh -s -k test_date_to_timestamp
```
```
!Expression <Cast> cast(a#4 as timestamp) 
cannot run on GPU because Timezone America/Los_Angeles is not supported yet. 
Only Non DST (daylight saving time) timezone is supported.
```

**Review by firestarman — APPROVED** [review]
Looks good to me but a NIT.

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/date_time_test.py [hunk 1]
```diff
@@ -671,3 +671,14 @@ def test_timestamp_millis_long_overflow():
 def test_timestamp_micros(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark : unary_op_df(spark, data_gen).selectExpr("timestamp_micros(a)"))
+
+
+@pytest.mark.parametrize('parser_policy', ["LEGACY", "CORRECTED", "EXCEPTION"], ids=idfn)
+def test_date_to_timestamp(parser_policy):
+    parser_policy_dic = {"spark.sql.legacy.timeParserPolicy": "{}".format(parser_policy)}
```
[the remark this answers is not in the document]
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py]: done

**Comment by firestarman** [comment]
premerge failed

File: integration_tests/src/main/python/date_time_test.py [hunk 2]
```diff
@@ -671,3 +671,14 @@ def test_timestamp_millis_long_overflow():
...
+
+@pytest.mark.parametrize('parser_policy', ["LEGACY", "CORRECTED", "EXCEPTION"], ids=idfn)
+def test_date_to_timestamp(parser_policy):
+    assert_gpu_and_cpu_are_equal_sql(
+        lambda spark : unary_op_df(spark, date_gen),
+        "tab",
+        "SELECT cast(a as timestamp) from tab",
+        conf = {
+            "spark.sql.legacy.timeParserPolicy": "{}".format(parser_policy),
+            "spark.rapids.sql.incompatibleDateFormats.enabled": True})
```
Comment (firestarman) [inline integration_tests/src/main/python/date_time_test.py:684]: Seems we only need `spark.rapids.sql.incompatibleDateFormats.enabled` for `LEGACY` policy?

**Comment by res-life** [comment]
build

File: integration_tests/src/main/python/date_time_test.py [hunk 3]
```diff
@@ -671,3 +671,14 @@ def test_timestamp_millis_long_overflow():
 def test_timestamp_micros(data_gen):
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark : unary_op_df(spark, data_gen).selectExpr("timestamp_micros(a)"))
+
+
+@pytest.mark.parametrize('parser_policy', ["LEGACY", "CORRECTED", "EXCEPTION"], ids=idfn)
```
Comment (gerashegalov) [inline integration_tests/src/main/python/date_time_test.py:676]: Any reason to use different quote marks? 
```suggestion
@pytest.mark.parametrize('parser_policy', ['LEGACY', 'CORRECTED', 'EXCEPTION'], ids=idfn)
```

File: integration_tests/src/main/python/date_time_test.py [hunk 4]
```diff
@@ -671,3 +671,14 @@ def test_timestamp_millis_long_overflow():
...
+def test_date_to_timestamp(parser_policy):
+    assert_gpu_and_cpu_are_equal_sql(
+        lambda spark : unary_op_df(spark, date_gen),
+        "tab",
+        "SELECT cast(a as timestamp) from tab",
+        conf = {
+            "spark.sql.legacy.timeParserPolicy": "{}".format(parser_policy),
```
Comment (gerashegalov) [inline integration_tests/src/main/python/date_time_test.py:683]: format redundant?
```suggestion
            "spark.sql.legacy.timeParserPolicy": parser_policy,
```

**Comment by res-life** [comment]
Spark 400 building failure is related to https://github.com/NVIDIA/spark-rapids/issues/11479

File: integration_tests/src/main/python/date_time_test.py (hunk 2 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py:684]: Yes.

File: integration_tests/src/main/python/date_time_test.py (hunk 3 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py:676]: Done

File: integration_tests/src/main/python/date_time_test.py (hunk 4 above)
Comment (res-life) [reply integration_tests/src/main/python/date_time_test.py:683]: Done.

**Comment by res-life** [comment]
build

**Review by firestarman — DISMISSED** [review]

**Comment by res-life** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Comment by res-life** [comment]
build

**Review by firestarman — APPROVED** [review]

**Pull request merged**